src/app/content/courses/create-your-sdk-with-codama/codama-from-scratch/en.mdx

```diff
@@ -0,0 +1,230 @@
+
+To create the foundation of your Codama IDL, we use `RootNode`. This node serves as the top-level container that holds your main `ProgramNode` as well as any additional programs that may be referenced by the main program.
+
+```ts
+const node = rootNode(programNode({ ... }));
+```
+
+### Program Node
+
+To define an entire on-chain program, we use `ProgramNode`. This node represents the complete program deployed on-chain and defines all the minimum viable elements such as accounts, instructions, PDAs, and errors.
+
+In addition to these core elements, this node accepts the program's name, version, deployment public key, and markdown documentation:
+
+
+```ts
+const node = programNode({
+    name: 'counter',
+    publicKey: '22222222222222222222222222222222222222222222',
+    version: '0.0.1',
+    docs: [],
+    accounts: [],
+    instructions: [],
+    definedTypes: [],
+    pdas: [],
+    errors: [],
+});
+```
+
+In the next section we'll examine all these elements in detail.
+
+### Account Node
+
+To define on-chain accounts, we use `AccountNode`. This node is characterized by its name, data structure, and optional attributes such as PDA definitions and account discriminators. 
+
+It represents the accounts that typically correspond to the `state.rs` file of your program.
+
+The docs field can be used to add documentation explaining what this account accomplishes within the program:
+
+```ts
+const node = accountNode({
+    name: 'token',
+    data: structTypeNode([
+        structFieldTypeNode({ name: 'mint', type: publicKeyTypeNode() }),
+        structFieldTypeNode({ name: 'owner', type: publicKeyTypeNode() }),
+        structFieldTypeNode({ name: 'amount', type: numberTypeNode('u64') }),
+    ]),
+    discriminators: [sizeDiscriminatorNode(72)],
+    size: 72,
+    pdas: [
+        pdaNode({
+            name: 'associated_token_account',
+            seeds: [
+                constantPdaSeedNodeFromString('utf8', 'token'),
+                variablePdaSeedNode('mint', publicKeyTypeNode()),
+                variablePdaSeedNode('owner', publicKeyTypeNode()),
+            ],
+        }),
+    ],
+});
+```
+
+### Instruction Node
+
+To define program instructions, we use `InstructionNode`. This node represents an instruction in a program and allows you to specify the discriminator, required accounts, and instruction data without complications.
+
+Additionally, you can include optional accounts. Based on your program's design, these optional accounts can be resolved in two ways: by being omitted or by passing the program ID as an account. You can select the resolution method using the `optionalAccountStrategy: "omitted" | "programId"` field.
+
+The docs field can be used to add documentation explaining what this instruction accomplishes within the program.
+
+```ts
+const node = instructionNode({
+    name: 'increment',
+    discriminators: [constantDiscriminatorNode(constantValueNode(numberTypeNode("u8"), numberValueNode(1)))],
```

Comment:
Probably best to use a `fieldDiscriminatorNode` in this example since there is a discriminator argument already.

```suggestion
    discriminators: [fieldDiscriminatorNode('discriminator')],
```